File: insight/frontend/src/features/lite-cut/state/projectCodec.js
import {
  LITE_CUT_ENCODERS,
  LITE_CUT_ENCODER_TIERS,
  LITE_CUT_CANVAS_FIT_VALUES,
  LITE_CUT_OUTPUT_DEFAULTS,
  LITE_CUT_OUTPUT_LIMITS,
  LITE_CUT_PROJECT_SCHEMA_VERSION,
  LITE_CUT_RANGE_MODES,
  LITE_CUT_TIMELINE_LIMITS,
} from "./projectContract.js";
import { AUDIO_MASTER_GAIN, clampAudioGain } from "../domain/audioContract.js";
import { reconcileTransitionEvents } from "./transitionModel.js";
import { normalizeVisualMaterialFields } from "../domain/visualMaterial.js";
import { normalizeTimelineTrackOrder } from "./timelineUtils.js";

const LEGACY_OUTPUT_FIELDS = ["frame_blend_enabled", "frame_blend_frames", "high_frame_downsample_enabled", "delivery_fps"];
const LEGACY_CLIP_FIELDS = ["transition_in", "transition_out", "canvas_fit"];
const LEGACY_OVERLAY_FIELDS = ["transition_in", "transition_out", "fade_in_sec", "fade_out_sec"];
const LEGACY_TEXT_FIELDS = ["anim_in", "anim_out"];

export class LiteCutProjectCompatibilityError extends Error {
  constructor(code, message) {
    super(message);
    this.name = "LiteCutProjectCompatibilityError";
    this.code = code;
  }
}

function assertCurrentProjectContract(body) {
  if (body?.schema_version !== LITE_CUT_PROJECT_SCHEMA_VERSION) {
    throw new LiteCutProjectCompatibilityError(
      "LITECUT_PROJECT_VERSION_UNSUPPORTED",
      `LiteCut project schema ${String(body?.schema_version)} is unsupported; schema ${LITE_CUT_PROJECT_SCHEMA_VERSION} is required`,
    );
  }
  const paths = [];
  for (const field of LEGACY_OUTPUT_FIELDS) {
    if (Object.prototype.hasOwnProperty.call(body?.output || {}, field)) paths.push(`output.${field}`);
  }
  const tracks = Array.isArray(body?.tracks) ? body.tracks : [];
  for (const [trackIndex, track] of tracks.entries()) {
    const clips = Array.isArray(track?.clips) ? track.clips : [];
    for (const [clipIndex, clip] of clips.entries()) {
      for (const field of LEGACY_CLIP_FIELDS) {
        if (Object.prototype.hasOwnProperty.call(clip || {}, field)) paths.push(`tracks[${trackIndex}].clips[${clipIndex}].${field}`);
      }
    }
  }
  const overlays = Array.isArray(body?.overlays) ? body.overlays : [];
  for (const [overlayIndex, overlay] of overlays.entries()) {
    for (const field of LEGACY_OVERLAY_FIELDS) {
      if (Object.prototype.hasOwnProperty.call(overlay || {}, field)) paths.push(`overlays[${overlayIndex}].${field}`);
    }
    for (const field of LEGACY_TEXT_FIELDS) {
      if (Object.prototype.hasOwnProperty.call(overlay?.text || {}, field)) paths.push(`overlays[${overlayIndex}].text.${field}`);
    }
  }
  if (paths.length) {
    throw new LiteCutProjectCompatibilityError(
      "LITECUT_LEGACY_PROJECT_FIELDS_UNSUPPORTED",
      `Retired LiteCut project fields are not supported: ${paths.sort().join(", ")}`,
    );
  }
}

/**
 * Normalize a current LiteCut project without reading or mutating application
 * state. Retired schemas and fields are rejected before any repair is applied.
 */
export function normalizeLiteCutBody(rawBody) {
  const source = rawBody == null ? { schema_version: LITE_CUT_PROJECT_SCHEMA_VERSION } : rawBody;
  if (!source || typeof source !== "object" || Array.isArray(source)) {
    throw new LiteCutProjectCompatibilityError("LITECUT_PROJECT_INVALID", "LiteCut project body must be an object");
  }
  assertCurrentProjectContract(source);
  const body = structuredClone(source);
  let changed = false;
  if (!Array.isArray(body.tracks)) {
    body.tracks = [];
    changed = true;
  } else {
    body.tracks = body.tracks.map((track) => {
      if (!track || typeof track !== "object" || typeof track.solo === "boolean") return track;
      changed = true;
      return { ...track, solo: false };
    });
  }
  if (!body.output || typeof body.output !== "object") {
    body.output = {
      dir: LITE_CUT_OUTPUT_DEFAULTS.dir,
      filename: LITE_CUT_OUTPUT_DEFAULTS.filename,
      width: LITE_CUT_OUTPUT_DEFAULTS.width,
      height: LITE_CUT_OUTPUT_DEFAULTS.height,
      fps: LITE_CUT_OUTPUT_DEFAULTS.fps,
      encoder: LITE_CUT_OUTPUT_DEFAULTS.encoder,
      encoder_tier: LITE_CUT_OUTPUT_DEFAULTS.encoder_tier,
      framemeld_enabled: LITE_CUT_OUTPUT_DEFAULTS.framemeld_enabled,
      canvas_fit: LITE_CUT_OUTPUT_DEFAULTS.canvas_fit,
      background_color: LITE_CUT_OUTPUT_DEFAULTS.background_color,
      blur_amount: LITE_CUT_OUTPUT_DEFAULTS.blur_amount,
      range_mode: LITE_CUT_OUTPUT_DEFAULTS.range_mode,
      range_start_sec: LITE_CUT_OUTPUT_DEFAULTS.range_start_sec,
      range_end_sec: LITE_CUT_OUTPUT_DEFAULTS.range_end_sec,
    };
    changed = true;
  } else {
    const outputDefaults = {
      width: [LITE_CUT_OUTPUT_DEFAULTS.width, LITE_CUT_OUTPUT_LIMITS.width.min, LITE_CUT_OUTPUT_LIMITS.width.max],
      height: [LITE_CUT_OUTPUT_DEFAULTS.height, LITE_CUT_OUTPUT_LIMITS.height.min, LITE_CUT_OUTPUT_LIMITS.height.max],
      fps: [LITE_CUT_OUTPUT_DEFAULTS.fps, LITE_CUT_OUTPUT_LIMITS.fps.min, LITE_CUT_OUTPUT_LIMITS.fps.max],
      blur_amount: [LITE_CUT_OUTPUT_LIMITS.blurAmount.default, LITE_CUT_OUTPUT_LIMITS.blurAmount.min, LITE_CUT_OUTPUT_LIMITS.blurAmount.max],
    };
    for (const [key, [fallback, minimum, maximum]] of Object.entries(outputDefaults)) {
      const raw = Number(body.output[key]);
      const next = Number.isInteger(raw) ? Math.max(minimum, Math.min(maximum, raw)) : fallback;
      if (raw !== next) {
        body.output[key] = next;
        changed = true;
      }
    }
    if (!LITE_CUT_CANVAS_FIT_VALUES.includes(body.output.canvas_fit)) {
      body.output.canvas_fit = LITE_CUT_OUTPUT_DEFAULTS.canvas_fit;
      changed = true;
    }
    const rawBackground = String(body.output.background_color || "").trim();
    const expandedBackground = /^#[0-9a-f]{3}$/i.test(rawBackground)
      ? `#${[...rawBackground.slice(1)].map((digit) => digit.repeat(2)).join("")}`
      : rawBackground;
    const backgroundColor = /^#[0-9a-f]{6}$/i.test(expandedBackground) ? expandedBackground.toLowerCase() : LITE_CUT_OUTPUT_DEFAULTS.background_color;
    if (body.output.background_color !== backgroundColor) {
      body.output.background_color = backgroundColor;
      changed = true;
    }
    if (!LITE_CUT_ENCODERS.includes(body.output.encoder)) {
      body.output.encoder = LITE_CUT_OUTPUT_DEFAULTS.encoder;
      changed = true;
    }
    if (!LITE_CUT_ENCODER_TIERS.includes(body.output.encoder_tier)) {
      body.output.encoder_tier = LITE_CUT_OUTPUT_DEFAULTS.encoder_tier;
      changed = true;
    }
    if (typeof body.output.framemeld_enabled !== "boolean") {
      body.output.framemeld_enabled = LITE_CUT_OUTPUT_DEFAULTS.framemeld_enabled;
      changed = true;
    }
    if (!LITE_CUT_RANGE_MODES.includes(body.output.range_mode)) {
      body.output.range_mode = LITE_CUT_OUTPUT_DEFAULTS.range_mode;
      changed = true;
    }
    const rawRangeStart = Number(body.output.range_start_sec);
    if (!Number.isFinite(rawRangeStart) || rawRangeStart < 0) {
      body.output.range_start_sec = LITE_CUT_OUTPUT_DEFAULTS.range_start_sec;
      changed = true;
    }
  }
  if (!body.tracks.some((t) => t?.type === "video")) {
    body.tracks.unshift({
      id: "v1",
      type: "video",
      label: "V1",
      locked: false,
      hidden: false,
      muted: false,
      solo: false,
      clips: [],
    });
    changed = true;
  }
  // Keep track labels deterministic when a current project is malformed.
  const videoTrackLabels = body.tracks.filter((t) => t?.type === "video").map((t) => String(t?.label || ""));
  if (videoTrackLabels.some((label) => !label) || new Set(videoTrackLabels).size !== videoTrackLabels.length) {
    let nextVideoLabel = 1;
    for (const track of body.tracks) {
      if (track?.type === "video") track.label = `V${nextVideoLabel++}`;
    }
    changed = true;
  }
  if (!body.tracks.some((t) => t?.type === "audio")) {
    body.tracks.push({
      id: "a1",
      type: "audio",
      label: "A1",
      locked: false,
      hidden: false,
      muted: false,
      solo: false,
      clips: [],
    });
    changed = true;
  }
  if (normalizeTimelineTrackOrder(body)) changed = true;
  if (!Array.isArray(body.overlays)) {
    body.overlays = [];
    changed = true;
  }
  if (!Array.isArray(body.markers)) {
    body.markers = [];
    changed = true;
  } else {
    body.markers = body.markers
      .map((m) => ({
        id: String(m?.id || `marker-${globalThis.crypto?.randomUUID?.()?.slice?.(0, 10) || Date.now()}`),
        time_sec: Math.max(LITE_CUT_TIMELINE_LIMITS.time.min, Math.min(LITE_CUT_TIMELINE_LIMITS.time.max, Number(m?.time_sec) || LITE_CUT_TIMELINE_LIMITS.time.default)),
        label: String(m?.label || ""),
        color: /^#[0-9a-f]{6}$/i.test(String(m?.color || "")) ? m.color : "#f59e0b",
      }))
      .sort((a, b) => a.time_sec - b.time_sec);
  }
  if (!body.audio || typeof body.audio !== "object") {
    body.audio = { master_volume: AUDIO_MASTER_GAIN.default };
    changed = true;
  } else {
    const raw = Number(body.audio.master_volume);
    if (!Number.isFinite(raw)) {
      body.audio.master_volume = AUDIO_MASTER_GAIN.default;
      changed = true;
    } else {
      const next = clampAudioGain(raw, AUDIO_MASTER_GAIN);
      if (next !== raw) {
        body.audio.master_volume = next;
        changed = true;
      }
    }
  }
  if (normalizeVisualMaterialFields(body)) changed = true;
  const transitionsBefore = JSON.stringify([body.transition_model_version, body.transitions]);
  reconcileTransitionEvents(body);
  if (JSON.stringify([body.transition_model_version, body.transitions]) !== transitionsBefore) changed = true;
  return { body, changed };
}

export function selectLiteCutProjectReferences(body) {
  const tracks = Array.isArray(body?.tracks) ? body.tracks : [];
  const overlays = Array.isArray(body?.overlays) ? body.overlays : [];
  const markers = Array.isArray(body?.markers) ? body.markers : [];
  const transitions = Array.isArray(body?.transitions) ? body.transitions : [];
  const clips = tracks.flatMap((track) => (Array.isArray(track?.clips) ? track.clips : []));
  return {
    trackIds: tracks.map((track) => String(track?.id || "")),
    clipIds: clips.map((clip) => String(clip?.id || "")),
    overlayIds: overlays.map((overlay) => String(overlay?.id || "")),
    markerIds: markers.map((marker) => String(marker?.id || "")),
    transitionIds: transitions.map((transition) => String(transition?.id || "")),
    sourceIds: clips
      .filter((clip) => clip?.source_id != null)
      .map((clip) => Number(clip.source_id))
      .filter(Number.isFinite),
  };
}

export function diagnoseLiteCutProjectReferences(body, { availableAssetIds = [] } = {}) {
  const diagnostics = [];
  const available = new Set((availableAssetIds || []).map(Number).filter(Number.isFinite));
  const references = selectLiteCutProjectReferences(body);
  const reportDuplicateIds = (kind, ids) => {
    const seen = new Set();
    for (const id of ids) {
      if (id && seen.has(id)) diagnostics.push({ code: `duplicate_${kind}_id`, kind, id });
      seen.add(id);
    }
  };
  reportDuplicateIds("track", references.trackIds);
  reportDuplicateIds("clip", references.clipIds);
  reportDuplicateIds("overlay", references.overlayIds);
  reportDuplicateIds("marker", references.markerIds);
  reportDuplicateIds("transition", references.transitionIds);

  for (const track of body?.tracks || []) {
    for (const clip of track?.clips || []) {
      if (clip?.source_type === "file" && !String(clip?.file_path || "").trim()) {
        diagnostics.push({ code: "missing_file_path", kind: "clip", id: String(clip?.id || "") });
      }
      if (clip?.source_id != null && !available.has(Number(clip.source_id))) {
        diagnostics.push({ code: "unresolved_source_id", kind: "clip", id: String(clip?.id || ""), source_id: Number(clip.source_id) });
      }
    }
  }
  for (const overlay of body?.overlays || []) {
    if (overlay?.type !== "text" && !String(overlay?.asset_path || "").trim()) {
      diagnostics.push({ code: "missing_overlay_asset_path", kind: "overlay", id: String(overlay?.id || "") });
    }
  }
  return diagnostics;
}

File: insight/backend/app/features/lite_cut/project_codec.py
"""Pure schema-v3 LiteCut project decoding, encoding, and diagnostics."""

from __future__ import annotations

import json
from dataclasses import asdict, dataclass
from functools import lru_cache
from typing import Any, Iterable

from importlib.resources.abc import Traversable

from .contracts import contract_resource
from .models import SCHEMA_VERSION, LiteCutProjectBody, empty_project
from .project_boundaries import normalized_project_boundaries
from .transition_events import normalized_transition_project
from .visual_material import normalized_visual_material_project


LEGACY_OUTPUT_FIELDS = frozenset({
    "frame_blend_enabled",
    "frame_blend_frames",
    "high_frame_downsample_enabled",
    "delivery_fps",
})
LEGACY_CLIP_FIELDS = frozenset({"transition_in", "transition_out", "canvas_fit"})
LEGACY_OVERLAY_FIELDS = frozenset({"transition_in", "transition_out", "fade_in_sec", "fade_out_sec"})
LEGACY_TEXT_FIELDS = frozenset({"anim_in", "anim_out"})


class LiteCutProjectCompatibilityError(ValueError):
    """Raised before normalization when a project uses a retired contract."""

    def __init__(self, code: str, message: str):
        super().__init__(message)
        self.code = code


def _assert_current_project_contract(raw: dict[str, Any]) -> None:
    actual_version = raw.get("schema_version")
    if actual_version != SCHEMA_VERSION:
        raise LiteCutProjectCompatibilityError(
            "LITECUT_PROJECT_VERSION_UNSUPPORTED",
            f"LiteCut project schema {actual_version!r} is unsupported; schema {SCHEMA_VERSION} is required",
        )

    legacy_paths: list[str] = []
    output = raw.get("output") if isinstance(raw.get("output"), dict) else {}
    legacy_paths.extend(f"output.{field}" for field in LEGACY_OUTPUT_FIELDS if field in output)
    for track_index, track in enumerate(raw.get("tracks") or []):
        if not isinstance(track, dict):
            continue
        for clip_index, clip in enumerate(track.get("clips") or []):
            if not isinstance(clip, dict):
                continue
            legacy_paths.extend(
                f"tracks[{track_index}].clips[{clip_index}].{field}"
                for field in LEGACY_CLIP_FIELDS
                if field in clip
            )
    for overlay_index, overlay in enumerate(raw.get("overlays") or []):
        if not isinstance(overlay, dict):
            continue
        legacy_paths.extend(
            f"overlays[{overlay_index}].{field}"
            for field in LEGACY_OVERLAY_FIELDS
            if field in overlay
        )
        text = overlay.get("text") if isinstance(overlay.get("text"), dict) else {}
        legacy_paths.extend(
            f"overlays[{overlay_index}].text.{field}"
            for field in LEGACY_TEXT_FIELDS
            if field in text
        )
    if legacy_paths:
        raise LiteCutProjectCompatibilityError(
            "LITECUT_LEGACY_PROJECT_FIELDS_UNSUPPORTED",
            f"Retired LiteCut project fields are not supported: {', '.join(sorted(legacy_paths))}",
        )


def project_contract_path() -> Traversable:
    return contract_resource("lite_cut_project_contract.json")


@lru_cache(maxsize=1)
def load_project_contract() -> dict[str, Any]:
    path = project_contract_path()
    try:
        payload = json.loads(path.read_text(encoding="utf-8"))
    except Exception as exc:
        raise RuntimeError(f"LiteCut project contract could not be loaded: {path}") from exc
    if not isinstance(payload, dict) or int(payload.get("project_schema_version") or 0) != SCHEMA_VERSION:
        raise RuntimeError(f"LiteCut project contract has an invalid schema: {path}")
    return payload


def create_empty_project() -> LiteCutProjectBody:
    return empty_project()


def validate_project_body(raw: dict[str, Any]) -> LiteCutProjectBody:
    _assert_current_project_contract(raw)
    bounded_body, _bounded_changed = normalized_project_boundaries(raw)
    visual_body, _visual_changed = normalized_visual_material_project(bounded_body)
    return LiteCutProjectBody.model_validate(normalized_transition_project(visual_body))


def read_project_body(raw: dict[str, Any] | None) -> LiteCutProjectBody:
    if raw is None:
        return create_empty_project()
    return validate_project_body(raw)


def serialize_project_body(project: LiteCutProjectBody | dict[str, Any] | None) -> dict[str, Any]:
    parsed = project if isinstance(project, LiteCutProjectBody) else read_project_body(project)
    return parsed.model_dump(mode="json", by_alias=True)


@dataclass(frozen=True)
class ProjectReferenceDiagnostic:
    code: str
    kind: str
    id: str = ""
    source_id: int | None = None

    def as_dict(self) -> dict[str, Any]:
        return {key: value for key, value in asdict(self).items() if value not in (None, "")}


def select_project_references(body: dict[str, Any] | None) -> dict[str, list[Any]]:
    source = body if isinstance(body, dict) else {}
    tracks = source.get("tracks") if isinstance(source.get("tracks"), list) else []
    overlays = source.get("overlays") if isinstance(source.get("overlays"), list) else []
    markers = source.get("markers") if isinstance(source.get("markers"), list) else []
    transitions = source.get("transitions") if isinstance(source.get("transitions"), list) else []
    clips = [clip for track in tracks if isinstance(track, dict) for clip in track.get("clips") or [] if isinstance(clip, dict)]
    return {
        "track_ids": [str(track.get("id") or "") for track in tracks if isinstance(track, dict)],
        "clip_ids": [str(clip.get("id") or "") for clip in clips],
        "overlay_ids": [str(overlay.get("id") or "") for overlay in overlays if isinstance(overlay, dict)],
        "marker_ids": [str(marker.get("id") or "") for marker in markers if isinstance(marker, dict)],
        "transition_ids": [str(transition.get("id") or "") for transition in transitions if isinstance(transition, dict)],
        "source_ids": [int(clip["source_id"]) for clip in clips if clip.get("source_id") is not None],
    }


def diagnose_project_references(
    body: dict[str, Any] | None,
    *,
    available_asset_ids: Iterable[int] = (),
) -> list[ProjectReferenceDiagnostic]:
    source = body if isinstance(body, dict) else {}
    diagnostics: list[ProjectReferenceDiagnostic] = []
    references = select_project_references(source)
    available = {int(value) for value in available_asset_ids}

    def report_duplicate_ids(kind: str, values: list[Any]) -> None:
        seen: set[str] = set()
        for raw in values:
            value = str(raw)
            if value and value in seen:
                diagnostics.append(ProjectReferenceDiagnostic(f"duplicate_{kind}_id", kind, value))
            seen.add(value)

    report_duplicate_ids("track", references["track_ids"])
    report_duplicate_ids("clip", references["clip_ids"])
    report_duplicate_ids("overlay", references["overlay_ids"])
    report_duplicate_ids("marker", references["marker_ids"])
    report_duplicate_ids("transition", references["transition_ids"])

    for track in source.get("tracks") or []:
        if not isinstance(track, dict):
            continue
        for clip in track.get("clips") or []:
            if not isinstance(clip, dict):
                continue
            clip_id = str(clip.get("id") or "")
            if clip.get("source_type") == "file" and not str(clip.get("file_path") or "").strip():
                diagnostics.append(ProjectReferenceDiagnostic("missing_file_path", "clip", clip_id))
            if clip.get("source_id") is not None and int(clip["source_id"]) not in available:
                diagnostics.append(ProjectReferenceDiagnostic("unresolved_source_id", "clip", clip_id, int(clip["source_id"])))
    for overlay in source.get("overlays") or []:
        if not isinstance(overlay, dict):
            continue
        if overlay.get("type") != "text" and not str(overlay.get("asset_path") or "").strip():
            diagnostics.append(ProjectReferenceDiagnostic("missing_overlay_asset_path", "overlay", str(overlay.get("id") or "")))
    return diagnostics


__all__ = [
    "ProjectReferenceDiagnostic",
    "LiteCutProjectCompatibilityError",
    "create_empty_project",
    "diagnose_project_references",
    "load_project_contract",
    "project_contract_path",
    "read_project_body",
    "select_project_references",
    "serialize_project_body",
    "validate_project_body",
]
